File: question4.cpp
#include <iostream>
#include <vector>
#include <set>
using namespace std;

vector<int> find_duplicates(vector<int>& nums) {
    vector<int> duplicates;
    set<int> seen;
    for (int num : nums) {
        if (seen.count(num)) {
            duplicates.push_back(num);
        } else {
            seen.insert(num);
        }
    }
    return duplicates;
}

int main() {
    vector<int> nums = {4, 3, 2, 7, 8, 2, 3, 1};
    vector<int> duplicates = find_duplicates(nums);

    cout << "Duplicates: ";
    for (int num : duplicates) {
        cout << num << " ";
    }
    cout << endl; // Output: 2 3

    return 0;
}

File: question4.py
def find_duplicates(nums):
    duplicates = []
    seen = set()
    for num in nums:
        if num in seen:
            duplicates.append(num)
        else:
            seen.add(num)
    return duplicates

# Example usage
nums = [4, 3, 2, 7, 8, 2, 3, 1]
print(find_duplicates(nums))  # Output: [2, 3]
